hazelcast/src/main/java/com/hazelcast/transaction/impl/xa/XAResourceImpl.java

```diff
@@ -0,0 +1,303 @@
+    public synchronized void end(Xid xid, int flags) throws XAException {
+        List<TransactionContext> contexts = getTransactionContexts(xid);
+        switch (flags) {
+            case TMFAIL:
+                for (TransactionContext context : contexts) {
+                    getTransaction(context).rollback();
+                }
+                break;
+            case TMSUCCESS:
+            case TMSUSPEND:
+                break;
+            default:
+                throw new XAException("Unknown flag!!! " + flags);
+        }
+    }
+
+    @Override
+    public synchronized int prepare(Xid xid) throws XAException {
+        List<TransactionContext> contexts = getTransactionContexts(xid);
+        for (TransactionContext context : contexts) {
+            Transaction transaction = getTransaction(context);
+            transaction.prepare();
+        }
+        return XA_OK;
+    }
+
+    @Override
+    public synchronized void commit(Xid xid, boolean onePhase) throws XAException {
+        List<TransactionContext> contexts = xidContextMap.remove(xid);
+        if (contexts == null && onePhase) {
+            throw new XAException("No context with the given xid: " + xid);
+        }
+        if (contexts == null) {
+            finalizeTransactionRemotely(xid, true);
+            return;
+        }
+
+        for (TransactionContext context : contexts) {
+            Transaction transaction = getTransaction(context);
+            if (onePhase) {
+                transaction.prepare();
+            }
+            transaction.commit();
+        }
+        clearRemoteTransactions(xid);
+    }
+
+    @Override
+    public synchronized void rollback(Xid xid) throws XAException {
+        List<TransactionContext> contexts = xidContextMap.remove(xid);
+        if (contexts == null) {
+            finalizeTransactionRemotely(xid, true);
+            return;
+        }
+        for (TransactionContext context : contexts) {
+            getTransaction(context).rollback();
+        }
+        clearRemoteTransactions(xid);
+    }
+
+    private void finalizeTransactionRemotely(Xid xid, boolean isCommit) throws XAException {
+        NodeEngine nodeEngine = getNodeEngine();
+        InternalPartitionService partitionService = nodeEngine.getPartitionService();
+        OperationService operationService = nodeEngine.getOperationService();
+        SerializableXID serializableXID =
+                new SerializableXID(xid.getFormatId(), xid.getGlobalTransactionId(), xid.getBranchQualifier());
+        Data xidData = nodeEngine.toData(serializableXID);
+        int partitionId = partitionService.getPartitionId(xidData);
+        FinalizeRemoteTransactionOperation operation = new FinalizeRemoteTransactionOperation(xidData, isCommit);
+        InternalCompletableFuture<Integer> future = operationService.invokeOnPartition(SERVICE_NAME, operation, partitionId);
+        Integer errorCode;
+        try {
+            errorCode = future.get();
```

Comment:
You're doing a blocking call under a `synchronized`block. Is it definitely required?